pallets/liquid-staking-two/src/lib.rs

```diff
@@ -0,0 +1,607 @@
+        who: &T::AccountId,
+        claim_era: &EraIndex,
+        current_era: &EraIndex,
+        user_ledger_per_era: &UserLedgerPerEra,
+        pool_ledger_per_era: &PoolLedgerPerEra,
+    ) -> WithdrawalAmount {
+        if claim_era.clone() + T::StakingDuration::get() > current_era.clone() {
+            if !user_ledger_per_era.claimed_matching {
+                MatchingQueueByUser::<T>::mutate(who, claim_era, |b| {
+                    b.claimed_matching = true;
+                });
+
+                user_ledger_per_era.instant_withdrawal_by_bond(pool_ledger_per_era)
+            } else {
+                (0, 0)
+            }
+        } else {
+            user_ledger_per_era.remaining_withdrawal_limit()
+        }
+    }
+
+    // if unbond, normally need to wait 28 eras
+    // after waiting 28 eras, users who unstake and get the left part
+    // considering users who stake and forget to claim within 28 eras.
+    fn claim_in_unbond_operation(
+        who: &T::AccountId,
+        claim_era: &EraIndex,
+        current_era: &EraIndex,
+        user_ledger_per_era: &UserLedgerPerEra,
+        pool_ledger_per_era: &PoolLedgerPerEra,
+    ) -> WithdrawalAmount {
+        if claim_era.clone() + T::BondingDuration::get() > current_era.clone() {
+            if !user_ledger_per_era.claimed_matching {
+                MatchingQueueByUser::<T>::mutate(who, claim_era, |b| {
+                    b.claimed_matching = true;
+                });
+
+                user_ledger_per_era.instant_withdrawal_by_unbond(pool_ledger_per_era)
+            } else {
+                (0, 0)
+            }
+        } else {
+            user_ledger_per_era.remaining_withdrawal_limit()
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    #[inline]
+    fn increase_staked_asset(amount: BalanceOf<T>) -> DispatchResult {
+        StakingPool::<T>::try_mutate(|m| -> DispatchResult {
+            *m = m.checked_add(amount).ok_or(ArithmeticError::Overflow)?;
+            Ok(())
+        })
+    }
+
+    #[inline]
+    fn decrease_staked_asset(amount: BalanceOf<T>) -> DispatchResult {
+        StakingPool::<T>::try_mutate(|m| -> DispatchResult {
+            *m = m.checked_sub(amount).ok_or(ArithmeticError::Underflow)?;
+            Ok(())
+        })
+    }
+
+    #[inline]
+    fn update_exchange_rate() -> DispatchResult {
+        let exchange_rate = Rate::checked_from_rational(
+            StakingPool::<T>::get(),
+            T::Currency::total_issuance(T::LiquidCurrency::get()),
+        )
+        .ok_or(Error::<T>::InvalidExchangeRate)?;
+        ExchangeRate::<T>::put(exchange_rate);
+        Ok(())
+    }
+
+    #[inline]
+    fn try_mark_op_succeed(
+        era_index: EraIndex,
+        op_type: StakingOperationType,
+    ) -> Result<BalanceOf<T>, DispatchError> {
+        StakingOperationHistory::<T>::try_mutate(
+            era_index,
+            op_type,
+            |op| -> Result<BalanceOf<T>, DispatchError> {
```

Comment:
`DispatchResult` should be preferred

Reply:
what's the different between `DispatchResult` and `Result`?

Reply:
no sorry DispatchResult's return value is `()` we cannot have `DispatchResult<T>`